src/Compilers/Core/Portable/Collections/ConcurrentCache.cs

```diff
@@ -37,7 +30,11 @@ internal Entry(int hash, TKey key, TValue value)
         }
 
         public ConcurrentCache(int size, IEqualityComparer<TKey> keyComparer)
-            : base(size)
+            // Defer creating the backing array until it is actually needed.  This saves on expensive allocations for
+            // short-lived compilations that do not end up using the cache.  As the cache is simple best-effort, it's
+            // fine if multiple threads end up creating the backing array at the same time.  One thread will be last and
+            // will win, and the others will just end up creating a small piece of garbage that will be collected.
```

Comment:
Feels like some version of this comment also needs to be on the `??=` in `CachingFactory` to explain why it's not using a threadsafe primitive.

Reply:
done.